src/Domain/CoordinateMaps/CylindricalEndcap.cpp

```diff
@@ -0,0 +1,224 @@
+
+#include <pup.h>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/CoordinateMaps/FocallyLiftedEndcap.hpp"
+#include "Domain/CoordinateMaps/FocallyLiftedMap.hpp"
+#include "Parallel/PupStlCpp11.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/DereferenceWrapper.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace domain::CoordinateMaps {
+
+CylindricalEndcap::CylindricalEndcap(const std::array<double, 3>& center_one,
+                                     const std::array<double, 3>& center_two,
+                                     const std::array<double, 3>& proj_center,
+                                     double radius_one, double radius_two,
+                                     double z_plane) noexcept
+    : impl_(center_two, proj_center, radius_two,
+            FocallyLiftedInnerMaps::Endcap(center_one, radius_one, z_plane)) {
+#ifdef SPECTRE_DEBUG
+  // There are two types of sanity checks here on the map parameters.
+  // 1) ASSERTS that guarantee that the map is invertible.
+  // 2) ASSERTS that guarantee that the map parameters fall within
+  //    the range tested by the unit tests (which is the range in which
+  //    the map is expected to be used).
+  //
+  // There are two reasons why 1) and 2) are not the same:
+  //
+  // a) It is possible to choose parameters such that the map is
+  //    invertible but the resulting geometry has very sharp angles,
+  //    very large or ill-conditioned Jacobians, or both.  We want to
+  //    avoid such cases.
+  // b) We do not want to waste effort testing the map for parameters
+  //    that we don't expect to be used.  For example, we demand
+  //    here that proj_center and sphere_one are contained within
+  //    sphere_two, but the map should still be valid for some choices
+  //    of parameters where sphere_one and sphere_two are disjoint;
+  //    allowing those parameter choices would involve much more
+  //    complicated logic to determine whether the map produces shapes
+  //    with sharp angles or large jacobians, and it would involve more
+  //    complicated unit tests to cover those possibilities.
+
+  // First test for invertibility.
+
+  // Consider the intersection of sphere_one and the plane formed by
+  // z_plane.  Call it circle_one.  Consider the cone with apex
+  // center_one that intersects sphere_one on circle_one. This cone
+  // has an opening angle 2*theta.  Call the cone 'cone_one'.
+  const double cos_theta = (z_plane - center_one[2]) / radius_one;
+
+  // Now consider a different cone, cone_new, constructed so that
+  // cone_new and cone_one intersect each other on circle_one at right
+  // angles.  The apex of cone_new is cone_new_apex, defined as
+  // follows:
+  const std::array<double, 3> cone_new_apex = {
+      center_one[0], center_one[1], center_one[2] + radius_one / cos_theta};
+  // Cone_new opens in the -z direction with opening angle 2*(pi/2-theta).
```

Comment:
If it's not too difficult to do, could you add a little coordinate axis symbol into the corner of the second diagram, "Allowed region for P" (or at least which way the z-axis is)